### PR TITLE
[pytorch] add manual registration for trace type

### PR DESCRIPTION
**Pull request by ljk53**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #41004 [pytorch] deprecate PYTORCH_DISABLE_TRACING macro
* **#40903 [pytorch] add manual registration for trace type**

This PR continues the work of #38467 - decoupling Autograd and Trace for manually registered ops.

Differential Revision: [D22354804](https://our.internmc.facebook.com/intern/diff/D22354804/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22354804/)!

### CODE REVIEW

File: test/jit/test_tracer.py [hunk 1]
```diff
@@ -765,6 +765,22 @@ def f(x):
         self.assertEqual(outputs, m(*inputs))
         self.assertExportImport(g, (x,))
 
+    def test_inplace_copy_force_outplace(self):
+        x = torch.randn(4, 4, requires_grad=True)
+
+        def f(x):
+            out = Variable(torch.zeros(x.size()))
```
Comment (albanD) [inline test/jit/test_tracer.py:772]: Wrapping into a Variable is a no-op. No need to do it here.

File: torch/csrc/autograd/TraceTypeManual.cpp [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    }
+    jit::tracer::ensureUniqueIfOutOfPlaced("copy_ (possibly due to an assignment)", self);
+  }
+#endif
+
+  static auto op = c10::Dispatcher::singleton()
+      .findSchemaOrThrow("aten::copy_", "")
+      .typed<Tensor & (Tensor &, const Tensor &, bool)>();
+  {
+    at::tracer::impl::NoTracerDispatchMode tracer_guard;
```
Comment (albanD) [inline torch/csrc/autograd/TraceTypeManual.cpp:43]: Out of curiosity, what is the reason we re-dispatch is done this way and not the same way it is done for VariableType which is:
```
  {
    at::AutoNonVariableTypeMode non_var_type_mode(true);
    self_.copy_(src_, non_blocking);
  }
```

Comment (ljk53) [reply torch/csrc/autograd/TraceTypeManual.cpp:43]: Short answer: yes, I probably should change to the more concise version here :)

More context:

There are 3 different ways to toggle tracing:
  1) c10 redispatch API. It dispatches to the lower level kernel (e.g. Tracer -> Autograd) by masking off the current dispatch key (Tracer). The mask only applies to this single call to avoid infinitely looping back to the Tracer kernel. It doesn't affect subsequent dispatches within the dispatch, i.e. op A Tracer --> op A CPU --> op B Tracer (it doesn't prevent it from dispatching to B's Tracer kernel).

  2) `NoTracerDispatchMode`. Similar to 1), except that it disables Tracer key for all subsequent dispatches within the scope.

  3) `jit::tracer::setTracingState(nullptr);` This is a domain switch to stop tracing.

We use different combinations of these switches at different places, which I plan to clean up next:

  - For generated TracerType kernels, we use 1) + 3) [Example](https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/autograd/generated/TraceType_0.cpp.html#34). 1) is a "domain-neural" way to do redispatching as in some cases we do want to call the same level kernel for other ops being called by the op, e.g. Profiler. The domain specific decision that we should skip all Tracing for intermediate ops is achieved by 3).

  - At some other places I used the `NoTracerDispatchMode` guard, usually it's where `AutoNonVariableTypeMode` was originally used. More details: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/TracerMode.h#L72

  - One difference between these manually registered kernels and generated kernels is that the original VariableTypeManual uses `AutoNonVariableTypeMode` instead of `setTracingState(null)` - `NoTracerDispatchMode` guard is inherited from there.

I guess I should choose 2) for this case for it's simplicity while thinking about what's the desired consistency.

Comment (albanD) [reply torch/csrc/autograd/TraceTypeManual.cpp:43]: Thanks for the detailed explanation!
So there is a crucial difference: The masking only effect the current dispatch but not the subsequent ones when using 1.
In the VariableType case, we actually never want this behavior: we either handle the autograd (and so the child ops should never do it) or not (and the child ops should always do it).

So for the tracer it is more complex as you might want to nest the tracing. Sounds good!

Comment (ljk53) [reply torch/csrc/autograd/TraceTypeManual.cpp:43]: @albanD yeah, it is exactly the key difference between redispatch v.s. guard.

For tracing specifically, it was originally inside VariableType so it automatically inherited no-nest-tracing behavior from the `AutoNonVariableTypeMode` guard. Besides, it also explicitly sets `setTracingState(nullptr)` before redispatching (except for these manually registered kernels which I'm not sure it's intentional or not). These two were overlapped but were added for different reasons.

So no-nest-tracing might actually be the invariant we want to stick with - as current tracing is for repeating the same behavior and tracing root ops are sufficient for this purpose. I simply haven't decided which pattern it should converge to. Now we have:
  * 1) + 3) for generated kernels;
  * 1) + 2) for these manually registered kernels; but 2) is totally a super set of 1) so I decided to use 2) only here as you questioned earlier;

Maybe changing everything to 2) is the right final state to move towards?

@ezyang @bhosmer - we talked about having a re-dispatching API that masks the dispatch-key for both current and subsequent calls, which is another option to converge to (if our goal is to reduce # of guards in the codebase :) )

**Comment by ezyang** [comment]
This is good stuff, but I think there's a decent chance the tracing logic here didn't need to be manually written out. Looking at the original manual code, you can see that there are bits that were codegenned, and then bits that were edited up to do some sort of custom logic. In most of the cases, the tracing logic was outside of the "this was manually edited sections." How hard would it be to wire up the codegen to generate tracing for those cases too?

(Maybe this is not worth it, since the PoR is to shortly delete all this code and replace it with boxed fallback.)

File: test/jit/test_tracer.py [hunk 3]
```diff
@@ -755,7 +755,7 @@ def test_inplace_copy(self):
         x = torch.randn(4, 4, requires_grad=True)
 
         def f(x):
-            out = Variable(torch.zeros(x.size()))
```
Comment (ezyang) [inline test/jit/test_tracer.py:758]: yay!

**Review by ezyang — APPROVED** [review]
There is not too much manual code that was carried over, so I think this is fine as is to go in.

**Comment by ljk53** [comment]
> There is not too much manual code that was carried over, so I think this is fine as is to go in.

Yeah, most ops here are simply falling through, among all the 5 manually registered ops: `copy_` has special logic to emit as `expand_as`, `resize_` and `resize_as_` only to emit warning message (not sure if it's still right thing to do?), only `detach` and `detach_` are regular form that can be replaced by codegen - but we'll need create a separate boolean flag to differentiate these 2 from other manual registered ops, so I decided not to do so...
